src/backend/cdb/dispatcher/cdbdisp_async.c

```
@@ -816,18 +816,22 @@ checkSegmentAlive(CdbDispatchCmdAsync *pParms)
 static inline void
 send_sequence_response(PGconn *conn, Oid oid, int64 last, int64 cached, int64 increment, bool overflow, bool error)
 {
-	pqPutMsgStart(SEQ_NEXTVAL_QUERY_RESPONSE, false, conn);
-	pqPutInt(oid, 4, conn);
-	pqPutInt(last >> 32, 4, conn);
-	pqPutInt(last, 4, conn);
-	pqPutInt(cached >> 32, 4, conn);
-	pqPutInt(cached, 4, conn);
-	pqPutInt(increment >> 32, 4, conn);
-	pqPutInt(increment, 4, conn);
-	pqPutc(overflow ? SEQ_NEXTVAL_TRUE : SEQ_NEXTVAL_FALSE, conn);
-	pqPutc(error ? SEQ_NEXTVAL_TRUE : SEQ_NEXTVAL_FALSE, conn);
-	pqPutMsgEnd(conn);
-	pqFlush(conn);
+	if (pqPutMsgStart(SEQ_NEXTVAL_QUERY_RESPONSE, false, conn) < 0 ||
+		pqPutInt(oid, 4, conn) < 0 ||
+		pqPutInt(last >> 32, 4, conn) < 0 ||
+		pqPutInt(last, 4, conn) < 0 ||
+		pqPutInt(cached >> 32, 4, conn) < 0 ||
+		pqPutInt(cached, 4, conn) < 0 ||
+		pqPutInt(increment >> 32, 4, conn) < 0 ||
+		pqPutInt(increment, 4, conn) < 0 ||
+		pqPutc(overflow ? SEQ_NEXTVAL_TRUE : SEQ_NEXTVAL_FALSE, conn) < 0 ||
+		pqPutc(error ? SEQ_NEXTVAL_TRUE : SEQ_NEXTVAL_FALSE, conn) < 0 ||
+		pqPutMsgEnd(conn) < 0 ||
+		pqFlush(conn) < 0)
+	{
+		Assert(conn->asyncStatus == PGASYNC_IDLE);
+		pqHandleSendFailure(conn);
+	}
```

Shall we report an error like "send sequence response failure", in case the receiver keep waiting for the response

---

Not sure whether should we raise an error here.  I checked other callers of the pqPut*() functions, most of them only call `pqHandleSendFailure()` on failure.  My understanding is that when pqPut*() returns `EOF` it probably indicates that the connection is closed, there is no chance to tell the receiver to give up, in fact the receiver should also check for `EOF` and handle the failure automatically.

---

Make sense.

---

@pengzhout is right, an error needs to be raised.  This code is executed in a QD backend.  An error raised here will cause the transaction to abort.  Which is the right thing to do if a QD backend couldn't send a response to nextval request from a QE.

---

Maybe we should find a way to trigger a failure here.  @dgkimura do you have any idea on how to reach here?

---

The `send_sequence_response()` function is called for a typical query involving sequence: e.g. insert into a table that has a serial column.